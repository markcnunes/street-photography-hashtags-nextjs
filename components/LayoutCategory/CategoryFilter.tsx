import React, { ChangeEvent, useState } from 'react';
import styled from '@emotion/styled';

export interface CategoryFilterProps {
  /**
   * Option of keywords available to be filtered.
   */
  options: string[];
  /**
   * Action to filter keywords.
   */
  filterKeywords: (arg: string[]) => void;
}

const CategoryFilter = ({ options, filterKeywords }: CategoryFilterProps) => {
  const [hidenCategories, setHidenCategories] = useState<string[]>([]);

  // Show or hide some specific category
  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const name = event.target.name;
    let updateHidenCategories;

    if (event.target.checked) {
      // add
      updateHidenCategories = [...hidenCategories, name];
    } else {
      // remove
      updateHidenCategories = hidenCategories.filter((item) => item !== name);
    }
    filterKeywords(updateHidenCategories);
    setHidenCategories(updateHidenCategories);
  };

  // Show or hide all keywords
  const handleInputChangeAll = (event: ChangeEvent<HTMLInputElement>) => {
    const name = event.target.name;
    let updateHidenCategories: string[];

    if (name === 'all') {
      updateHidenCategories = [];
    } else {
      // if name = none
      updateHidenCategories = options;
    }

    filterKeywords(updateHidenCategories);
    setHidenCategories(updateHidenCategories);
  };

  return (
    <StyledCategoryFilter>
      <h4>Filter:</h4>
      <label
        className={
          hidenCategories.length === options.length ? 'active' : undefined
        }
      >
        Show All
        <input
          name='all'
          type='checkbox'
          checked={hidenCategories.length === options.length}
          onChange={handleInputChangeAll}
        />
      </label>
      <label className={hidenCategories.length === 0 ? 'active' : undefined}>
        Hide All
        <input
          name='none'
          type='checkbox'
          checked={hidenCategories.length === 0}
          onChange={handleInputChangeAll}
        />
      </label>
      {options.map((item) => (
        <label
          key={item}
          className={
            hidenCategories.some((category) => category === item)
              ? 'active'
              : undefined
          }
        >
          {item}
          <input
            name={item}
            type='checkbox'
            checked={hidenCategories.some((category) => category === item)}
            onChange={handleInputChange}
          />
        </label>
      ))}
    </StyledCategoryFilter>
  );
};

export default CategoryFilter;

const StyledCategoryFilter = styled('form')(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  flexWrap: 'wrap',
  width: '100%',
  padding: 10,
  fontSize: theme.fontSizes[1],
  color: theme.colors.white,
  border: `1px solid ${theme.colors.white}`,
  borderRadius: 5,
  marginBottom: 20,
  h4: {
    margin: 0,
    width: '100%',
    textAlign: 'center',
  },
  label: {
    display: 'flex',
    alignItems: 'center',
    padding: 15,
    lineHeight: 0,
    cursor: 'pointer',
    borderRadius: 3,
    margin: 5,
    textTransform: 'capitalize',
    '&.active': {
      opacity: 0.4,
    },
    '&:hover': {
      color: theme.colors.black,
      background: theme.colors.white,
    },
    input: {
      display: 'none',
    },
  },
  '@media (min-width: 780px)': {
    h4: {
      width: 'auto',
      lineHeight: 1,
      marginRight: 10,
    },
  },
}));
