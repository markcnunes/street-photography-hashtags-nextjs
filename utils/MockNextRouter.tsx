import { RouterContext } from 'next/dist/next-server/lib/router-context';
import { NextRouter } from 'next/router';
import { ReactNode } from 'react';

interface MockNextRouterProps extends Partial<NextRouter> {
  children?: ReactNode;
}

/**
 * Override the default NextRouter render with different attributes.
 * e.g. change the path or locale for tests.
 */

const MockNextRouter = ({ children, ...props }: MockNextRouterProps) => {
  const mockRouter: NextRouter = {
    route: '/',
    pathname: '',
    query: {},
    asPath: '/',
    basePath: '',
    locale: 'en',
    locales: [''],
    defaultLocale: 'en',
    // domainLocales?: DomainLocales,
    isLocaleDomain: true,
    push: jest.fn(() => Promise.resolve(true)),
    replace: jest.fn(() => Promise.resolve(true)),
    reload: jest.fn(() => Promise.resolve(true)),
    back: jest.fn(() => Promise.resolve(true)),
    prefetch: jest.fn(),
    beforePopState: jest.fn(() => Promise.resolve(true)),
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
    isFallback: false,
    isReady: true,
    isPreview: false,
  };

  return (
    <RouterContext.Provider value={{ ...mockRouter, ...props }}>
      {children}
    </RouterContext.Provider>
  );
};

export default MockNextRouter;
