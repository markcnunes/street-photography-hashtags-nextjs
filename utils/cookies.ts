// import ReactGA from 'react-ga';

import GA4React from 'ga-4-react';

export const cookieConsent = 'sphCookieconsent';

/*
 * Check if cookie consentment is true before initializing and rendering it
 */
export const getCookie = (key: string): string | undefined => {
  var b = document.cookie.match('(^|;)\\s*' + key + '\\s*=\\s*([^;]+)');
  return b ? b.pop() : '';
};

export const gaCode = 'G-W7QKGLQ9YV';
const initializeCookies = () => {
  if (!window.location.href.includes('localhost')) {
    const ga4react = new GA4React(gaCode);
    ga4react.initialize();
    // reload page to get the track running
    window.location.reload();
  }
};

export const renderCookies = (): any => {
  getCookie(cookieConsent) === 'true' && initializeCookies();
};

export default initializeCookies;
