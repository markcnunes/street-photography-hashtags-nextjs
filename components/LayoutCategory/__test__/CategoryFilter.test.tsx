import userEvent from '@testing-library/user-event';
import React from 'react';
import { render, screen } from 'utils/test-utils';
import CategoryFilter from '../CategoryFilter';

const mockOptions = ['first', 'second', 'third'];

const mockFilterKeywords = jest.fn();

describe('<CategoryFilter />', () => {
  it('matchs snapshot', () => {
    const { asFragment } = render(
      <CategoryFilter
        options={mockOptions}
        filterKeywords={mockFilterKeywords}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders whiout props', () => {
    const { asFragment } = render(
      <CategoryFilter options={[]} filterKeywords={mockFilterKeywords} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders props', () => {
    render(
      <CategoryFilter
        options={mockOptions}
        filterKeywords={mockFilterKeywords}
      />,
    );
    expect(screen.getByText(mockOptions[0])).toBeInTheDocument();
  });
  it('hide and show category', () => {
    const { container } = render(
      <CategoryFilter
        options={mockOptions}
        filterKeywords={mockFilterKeywords}
      />,
    );
    const input = screen.getByText(mockOptions[0]);
    const active = container.getElementsByClassName('active');

    // hide category
    userEvent.click(input);

    expect(active.length).toBe(1);
    expect(input).toHaveStyle({
      opacity: 0.4,
    });

    // show category
    userEvent.click(active[0]);
    expect(active.length).toBe(1); // stills 1 because show or hide all buttons inherit it
    expect(input).not.toHaveStyle({
      opacity: 0.4,
    });
  });
});
