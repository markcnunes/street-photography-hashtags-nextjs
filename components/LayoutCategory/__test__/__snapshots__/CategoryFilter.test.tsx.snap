// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryFilter /> matchs snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  padding: 10px;
  font-size: 14px;
  color: #fff;
  border: 1px solid #fff;
  border-radius: 5px;
  margin-bottom: 20px;
}

.emotion-0 h4 {
  margin: 0;
  width: 100%;
  text-align: center;
}

.emotion-0 label {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px;
  line-height: 0;
  cursor: pointer;
  border-radius: 3px;
  margin: 5px;
  text-transform: capitalize;
}

.emotion-0 label.active {
  opacity: 0.4;
}

.emotion-0 label:hover {
  color: #1c1c27;
  background: #fff;
}

.emotion-0 label input {
  display: none;
}

@media (min-width: 780px) {
  .emotion-0 h4 {
    width: auto;
    line-height: 1;
    margin-right: 10px;
  }
}

<form
    class="emotion-0 emotion-1"
  >
    <h4>
      Filter:
    </h4>
    <label>
      Show All
      <input
        name="all"
        type="checkbox"
      />
    </label>
    <label
      class="active"
    >
      Hide All
      <input
        checked=""
        name="none"
        type="checkbox"
      />
    </label>
    <label>
      first
      <input
        name="first"
        type="checkbox"
      />
    </label>
    <label>
      second
      <input
        name="second"
        type="checkbox"
      />
    </label>
    <label>
      third
      <input
        name="third"
        type="checkbox"
      />
    </label>
  </form>
</DocumentFragment>
`;

exports[`<CategoryFilter /> renders whiout props 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  padding: 10px;
  font-size: 14px;
  color: #fff;
  border: 1px solid #fff;
  border-radius: 5px;
  margin-bottom: 20px;
}

.emotion-0 h4 {
  margin: 0;
  width: 100%;
  text-align: center;
}

.emotion-0 label {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px;
  line-height: 0;
  cursor: pointer;
  border-radius: 3px;
  margin: 5px;
  text-transform: capitalize;
}

.emotion-0 label.active {
  opacity: 0.4;
}

.emotion-0 label:hover {
  color: #1c1c27;
  background: #fff;
}

.emotion-0 label input {
  display: none;
}

@media (min-width: 780px) {
  .emotion-0 h4 {
    width: auto;
    line-height: 1;
    margin-right: 10px;
  }
}

<form
    class="emotion-0 emotion-1"
  >
    <h4>
      Filter:
    </h4>
    <label
      class="active"
    >
      Show All
      <input
        checked=""
        name="all"
        type="checkbox"
      />
    </label>
    <label
      class="active"
    >
      Hide All
      <input
        checked=""
        name="none"
        type="checkbox"
      />
    </label>
  </form>
</DocumentFragment>
`;
